src/commutator_rpnl.F

```diff
@@ -398,4 +402,509 @@ SUBROUTINE build_com_rpnl(matrix_rv, qs_kind_set, sab_orb, sap_ppnl, eps_ppnl)
+      TYPE(gth_potential_type), POINTER                  :: gth_potential
+      TYPE(gto_basis_set_p_type), DIMENSION(:), POINTER  :: basis_set
+      TYPE(gto_basis_set_type), POINTER                  :: orb_basis_set
+      TYPE(neighbor_list_iterator_p_type), &
+         DIMENSION(:), POINTER                           :: nl_iterator
+      TYPE(sap_int_type), DIMENSION(:), POINTER          :: sap_int
+      TYPE(sgp_potential_p_type), DIMENSION(:), POINTER  :: spotential
+      TYPE(sgp_potential_type), POINTER                  :: sgp_potential
+
+      CALL timeset(routineN, handle)
+
+      my_rxrv = .FALSE.
+      my_rrv = .FALSE.
+      my_rv = .FALSE.
+      IF (PRESENT(matrix_rxrv)) my_rxrv = .TRUE.
+      IF (PRESENT(matrix_rrv)) my_rrv = .TRUE.
+      IF (PRESENT(matrix_rv)) my_rv = .TRUE.
+      IF (.NOT. (my_rv .OR. my_rxrv .OR. my_rrv)) RETURN
+
+      ppnl_present = ASSOCIATED(sap_ppnl)
+
+      IF (ppnl_present) THEN
+
+         nkind = SIZE(qs_kind_set)
+
+         CALL get_qs_kind_set(qs_kind_set, &
+                              maxco=maxco, &
+                              maxlgto=maxlgto, &
+                              maxsgf=maxsgf, &
+                              maxlppnl=maxlppnl, &
+                              maxppnl=maxppnl)
+
+         maxl = MAX(maxlgto, maxlppnl)
+         CALL init_orbital_pointers(maxl + 1)
+
+         ldsab = MAX(maxco, ncoset(maxlppnl), maxsgf, maxppnl)
+         ldai = ncoset(maxl + 1)
+
+         !sap_int needs to be shared as multiple threads need to access this
+         ALLOCATE (sap_int(nkind*nkind))
+         DO i = 1, nkind*nkind
+            NULLIFY (sap_int(i)%alist, sap_int(i)%asort, sap_int(i)%aindex)
+            sap_int(i)%nalist = 0
+         END DO
+
+         !set up direct access to basis and potential
+         ALLOCATE (basis_set(nkind), gpotential(nkind), spotential(nkind))
+         DO ikind = 1, nkind
+            CALL get_qs_kind(qs_kind_set(ikind), basis_set=orb_basis_set)
+            IF (ASSOCIATED(orb_basis_set)) THEN
+               basis_set(ikind)%gto_basis_set => orb_basis_set
+            ELSE
+               NULLIFY (basis_set(ikind)%gto_basis_set)
+            END IF
+            CALL get_qs_kind(qs_kind_set(ikind), gth_potential=gth_potential, &
+                             sgp_potential=sgp_potential)
+            IF (ASSOCIATED(gth_potential)) THEN
+               gpotential(ikind)%gth_potential => gth_potential
+               NULLIFY (spotential(ikind)%sgp_potential)
+            ELSE IF (ASSOCIATED(sgp_potential)) THEN
+               spotential(ikind)%sgp_potential => sgp_potential
+               NULLIFY (gpotential(ikind)%gth_potential)
+            ELSE
+               NULLIFY (gpotential(ikind)%gth_potential)
+               NULLIFY (spotential(ikind)%sgp_potential)
+            END IF
+         END DO
+
+         IF (my_rxrv .OR. my_rrv) THEN
+            maxder = 10
+            order = 2
+            CPASSERT(PRESENT(ref_point))
```

Comment:
I'd shift this precondition check up to the others (together with the one further down below). And possible add some comment behind the `CPASSERT`s.